src/posts/2020-07-27-unsucking-your-project-onboarding.md

```diff
@@ -0,0 +1,789 @@
+internal project we had to set up `ngrok` before starting the server. Because
+this is something we tend to forget and newcommers might not be aware of, I made
+this `bin/` script that launches `ngrok` in the background, gets generated URL,
+launches the server and kills `ngrok` before exiting.
+
+```shell
+#!/usr/bin/env bash
+
+trap ctrl_c INT
+
+ctrl_c() {
+  printf "\n"
+  echo "Received kill signal, stopping ngrok... "
+  $(killall ngrok &>/dev/null);
+
+  printf "Done!\n"
+
+  exit 1
+}
+
+PORT=${PORT:-4000}
+
+echo "Starting ngrok... "
+
+ngrok http $PORT > /dev/null &
+
+sleep 2
+
+REGEX="\"public_url\":\s*\"(https://([^\"]+))\"[^{]+{\"addr\":\s*\"http:\/\/localhost:$PORT\""
+
+JSON_RESPONSE=$(curl http://localhost:4040/api/tunnels 2>/dev/null)
+
+if [[ "$JSON_RESPONSE" =~ $REGEX ]]; then
+  export WEBHOOK_URL=${BASH_REMATCH[1]}
+  export HOST=${BASH_REMATCH[2]}
+  export ENDPOINT_PORT="443"
+  export SCHEME="https"
+
+  printf "Done!\n\n"
+  echo "Your webhook url is $WEBHOOK_URL"
+else
+  echo "Wow, something went wrong with ngrok!";
+  ctrl_c
+fi
+
+echo "Starting development server with options HOST=$HOST SCHEME=$SCHEME ENDPOINT_PORT=$ENDPOINT_PORT..."
+
+iex -S mix
+```
+
+<a href="https://asciinema.org/a/Jbbe7Fvt5OYHEvqAfqEAyJeXc" target="_blank">
+  <img src="https://asciinema.org/a/Jbbe7Fvt5OYHEvqAfqEAyJeXc.svg" alt="bin/server in action" style="max-width: 100%"/>
+</a>
+
+Anyone coming into the project doesn't need to be aware of this complexity, they
+would just run a command and voilá. Magic.
+
+### bin/lint
+
+Usually nothing more than the scripts you would run on CI. As an example, here's
+what we have on a project that runs the linters for frontend (in JavaScript) and
+backend (in Elixir). Notice the `set -e`. This causes the script to exit if a
+command has a non-zero exit code. The reasoning behind this is to mimic the CI
+behaviour.
+
+```shell
+#!/usr/bin/env sh
+
+set -e
+
+echo "running the CSS linter..."
+yarn lint-styles
+
+echo "running the javascript linter..."
+yarn lint
+
+echo "running the elixir formatter..."
+mix format --check-formatted
+
+echo "running elixir credo..."
+mix credo --strict
+
+echo "no problems found!"
+```
+
+<a href="https://asciinema.org/a/5LcwnppxHGFAh4AViIMYG9gme" target="_blank">
+  <img src="https://asciinema.org/a/5LcwnppxHGFAh4AViIMYG9gme.svg" alt="bin/server in action" style="max-width: 100%"/>
+</a>
+
+### bin/setup
+
+Ah, the crown jewel. Usually the most complex script as it installs everything
+a developer might need. We usually make a few small assumptions:
+
+1. You have `postgres` installed. It's the database we use in almost every
+   project so 99.9% of the time we have it available in the system. Those 0.1%
+   of the time correspond to open source contributors or when we have a fresh
+   install (but we'll get into that). It's difficult to guess the way someone
+   might install `postgres` and for such a rare edge case, the script would get
+   unnecessarily complex.
```

Comment:
to be fair, asdf-postgres is a thing, so you may want to suggest it here, instead of just dismissing with "postgres is difficult to install". In our case I agree with the 99.9% rule, but that's *our* case